### ISSUE TITLE
Update `the_content` with the correct image format

### ISSUE THREAD

**Issue by mitogh** [WordPress/performance#149]
## Feature Description

Once an image attachment has more than one type of image available we need to make sure the correct images are rendered:

## Scenario 1 - Render WebP by default

- If the requested browser support WebP continue with the flow, otherwise render a JPEG version.
- If the attachment has a WebP version of the image replace any JPEG version with the exact WebP version of the image. 
- If no WebP version exists keep serving the JPEG version

## Scenario 2 - Render JPEG version

- If the browser does not accepts WebP version render a JPEG version
- if the current attachment does not have an exact WebP version for the requested image deliver the JPEG version
- If the WebP is disabled via a filter keep rendering JPEG by default.

This filter would be placed in `the_content` so when the article or articles are rendered the delivered content to the users uses the appropriate image format.

This work is a subsequent step for:

- #142

**Comment by felixarntz** [comment]
@mitogh 
> If the requested browser support WebP continue with the flow

I thought the idea was to always load WebP in the frontend (if the image exists in that version), based on the previous discussions on the Slack chat and the decision made in https://github.com/WordPress/performance/issues/96#issuecomment-1027147051

We decided to keep generating JPEG versions not for the frontend, but for other usages that are more niche but relevant, e.g. email newsletters.

In other words, I don't think we need to worry about scenario 2 here - we can rely on WordPress core's definition of officially supported browsers, all of which support WebP.

**Comment by mitogh** [comment]
Correct @felixarntz , the main idea would be to render WebP by default, if the browser supports it which is the majority of browsers at this point but for the cases where the browser does not support WebP the JPEG version would be displayed instead. By doing this we allow displaying the image at all times for older and more modern browsers, and this does not require a lot of extra work as is just checking against the `HTTP_ACCEPT` header which I think is a great benefit so modern and older browsers can still work as they are where modern browsers would take the WebP version by default (as WebP is supported there).

For example, if we load WebP by default this is how it would look in Safari when this validation is not present.

![2022-02-04_08-07](https://user-images.githubusercontent.com/3921289/152543845-593a23ec-9350-42fe-8baa-c4f5a49fc3ac.png)

And this is how it looks when the validation is present, the JPEG version is loaded instead.

![2022-02-03_19-28](https://user-images.githubusercontent.com/3921289/152544434-bca175ce-da3b-49a3-9b49-9ccd92beb809.png)

And for modern browsers, a WebP version is loaded instead.

![2022-02-03_19-28_1](https://user-images.githubusercontent.com/3921289/152544538-64c2cfdc-9b19-4a03-b982-cac8648148bd.png)

One of the major problems with this approach is the fact when using cache mechanics Which stores the full HTML of the page we won't able to detect this header from the client (Browser) so in that case, WebP would be loaded all the time, but a filter would be in place as described in https://github.com/WordPress/performance/issues/96#issuecomment-1027147051 to change this behavior.

**Comment by adamsilverstein** [comment]
Note that the intention in in #142 will be to store the default 'sizes' as WebP, so these would automatically be used for the `srcset` attributes. The one exception is the "full" sized image referenced in the `src` attribute, we need to ensure this uses the WebP file as well, currently this still references the original uploaded jpeg or -scaled version of the jpeg.

I'm also wondering whether an output mime type filter makes sense, since we plan to store WebP and jpeg versions, it should be possible to switch the default output type at any time (as long as the files are available).

**Comment by mitogh** [comment]
Created a PR with the implementation in place to replace `the_content`

- https://github.com/WordPress/performance/pull/152

**Comment by felixarntz** [comment]
Thanks @mitogh for the clarification. Overall what you're suggesting makes sense to me, relying on the `Accept` HTTP header. However, as you're already saying, this will have a problem with cached content. How will we be able to reliably know whether a request to the server will later be cached or not? Let me elaborate:

* Imagine a request comes through to WordPress from a browser that doesn't support WebP per its `Accept` header.
* WP will recognize that and not serve WebP.
* Since the site has a caching layer, the content is now cached, serving JPEG to all sites at least until that cache is invalidated.

The above would not be what we want, so how can we prevent that?

**Comment by mitogh** [comment]
Great point @felixarntz  so the idea would be to add a filter inside of the plugin like this:  

```php
add_filter( 'wp_preferred_image_mime', function() {
     return 'image/webp';
} );
```

Applied here https://github.com/WordPress/performance/pull/152/files#diff-71ed188a16b79d179868ee80e76f775462a451e871fc3d6214fcacca3edcea8cR356 that would deliver WebP all the time in the content.

**Comment by felixarntz** [comment]
@mitogh I understand the code, but as far as I can tell it doesn't solve the problem from the scenario I mentioned above. Could you please double-check?

**Comment by mitogh** [comment]
Right in this scenario, we have 2 options: 

1. Always deliver an image format as described in https://github.com/WordPress/performance/issues/96#issuecomment-1027147051 which can be accomplished with https://github.com/WordPress/performance/issues/149#issuecomment-1030212690 so that would deliver WebP all the time, which would be the default behavior in the plugin, deliver WebP regardless. This would be used for the scenario you described so even if the page is cached the WebP image would be cached as part of the markup.
2. Another path is to leave this dynamic removing this filter https://github.com/WordPress/performance/issues/149#issuecomment-1030212690 this would deliver the best possible image when the page is delivered. 

I think those 2 cases cover the majority of use cases:

- Enable WebP by default
- Use WebP if full HTML cache is in place
- Dynamically determine the best image format when Cache is not present
- Delivery only a JPEG instead of WebP

In regards to your specific use case described https://github.com/WordPress/performance/issues/149#issuecomment-1030192647 would fall into _Use WebP if full HTML cache is in place_ at least not from the server. 

We could introduce a JS piece of code to review if the client supports WebP if the client does not support WebP we replace the src of the images with a JPEG version in order to be compatible as much as possible with the clients. 

Let me know what do you think about this approach so I can explore some alternatives for this by enqueueing an inline JS to see how we can make this happen. Let me know in case you have additional questions. @felixarntz

**Comment by felixarntz** [comment]
@mitogh I see, so you're saying that developers would need to use the filter to enforce WebP to show on sites with page caching, to avoid JPEG being served just because whichever request caused the cache to be populated was from a browser that didn't support WebP, correct?

In that case, I don't think that would be a solution we should strive for, since it would require developer intervention to serve WebP in a use-case that is quite common. Caching layers are very common on WordPress sites, and unfortunately we cannot control them from WP core.

That's why I'm still leaning towards using WebP unconditionally by default, since the `Accept` header doesn't come through caching and therefore is unreliable. Is the primary concern about doing that Safari?

**Comment by mitogh** [comment]
Not necessarily, the behavior is as follows:

WebP is enabled and delivered by default, from there there are 2 levels of granularity that you can use to control this: 

1. Leave the decision to the client request
2. Force JPEG using the filter from https://github.com/WordPress/performance/issues/149#issuecomment-1030212690 

But WebP would be delivered and selected by default regardless of the value of `Accept`.

In addition to that, I feel like a client solution might be a good option if we are looking to maximize compatibility during the delivery of the images. @felixarntz

**Comment by felixarntz** [comment]
> Leave the decision to the client request

That's precisely my concern here, we can't really do that, at least not using the `Accept` HTTP header since it's unreliable. Again, why do we need to do that? Is it because of Safari? So far from my understanding, I thought we would be able to unconditonally deliver WebP, given WordPress core's browser support policy, but I may be wrong. Can you clarify why you think we need to check for support from the client?

I'm leaning against using the `Accept` header for the above reasons. If we decide that we need to check for support, a client-side mechanism may be the only reliable way - or the `picture` element.

**Comment by mitogh** [comment]
> That's precisely my concern here, we can't really do that,

Right, this is not the default behavior would be to deliver a WebP version regardless, however in order to expand the range of places where WebP is not supported, this mechanism would give the chance to a developer to unhook and allow to render JPEG in those cases, with the limitations of that system being invalid when a cache system is in place. 

>  Can you clarify why you think we need to check for support from the client?

Sure, the main reason for this is to make sure that an image is presented to the client regardless of the support and make sure the image is delivered.

**Comment by felixarntz** [comment]
I'm still leaning towards not using the `Accept` header. I understand what you're saying, but we shouldn't implement an unreliable detection mechanism in core and then put responsibility on developers to use a filter to fix it for their use-case. Let's take a step back here:
* This PR is, most importantly, for implementing the content replacement.
* I suggest for now (as in this PR) we just use WebP at all times, as long as the image exists in that format. That allows us to really focus on that aspect here.
* Regarding the client support detection, this needs more exploration, we need to discuss alternate approaches and decide on one. Working on code for this already without having made a decision is a bit too early now.

To be clear, I agree with you we need to explore the best ways to potentially support clients that don't support WebP here. But this hasn't been discussed and defined yet, and it is another layer of complexity. I think it's easier to have multiple iterative PRs with a very focused purpose. We can first merge the basic one, and then eventually work on a follow-up PR to improve the behavior for clients without WebP support.

**Comment by mitogh** [comment]
Thanks for the feedback @felixarntz and I do agree with those points let's break down those steps as actionable issues and remove the `Accept` header for now and iterate with the idea to deliver the best image format when possible but for now, let's just deliver WebP by default with the filters to disable it and render JPEG in those cases.

**Comment by adamsilverstein** [comment]
We would only update content for images inserted from the media library, right? 

Do we really need to filter the_content, or can we adjust images output by `wp_get_attachment_image` using the `wp_get_attachment_image` filter? That way we could avoid using a regex :)

**Comment by mitogh** [comment]
> We would only update content for images inserted from the media library, right?

Correct.

> Do we really need to filter the_content, or can we adjust images output by wp_get_attachment_image using the wp_get_attachment_image filter? That way we could avoid using a regex :)

Yes, this would be required for all existing content that was published before this work is released, unless this text transformation is directly applied into the DB which would be my preferred approach as this would require only 1 pass and the filter won't be required when delivering the content from the DB to the client.

Basically, we would need a task like:

- #24 

Where all the JPEG references to media images are updated in the DB to WebP images if they exist, otherwise we have to perform a filter on `the_content` every time the article is delivered.

**Comment by felixarntz** [comment]
I personally prefer the PHP replacement, since it allows for more flexibility and works out of the box, without requiring an extensive routine, which would put more responsibility on the site owner and take a lot more work to implement.

I would especially suggest this route since such regexes to parse images out of `the_content` already exist (see [`wp_filter_content_tags()`](https://developer.wordpress.org/reference/functions/wp_filter_content_tags/)). We could potentially tap into that for the logic here, although that would require adding one more filter in core for that. In any case, the final WP core implementation could rely on that, so we would not need to add yet another regex. So I think the additional cost here would be fairly low, especially since image metadata is already being loaded as well in any case, to populate the `srcset` and `sizes` attributes, which also happens dynamically.

**Comment by mitogh** [comment]
Hey @eugene-manuilov just a heads up this one already has a PR open: 

- https://github.com/WordPress/performance/pull/152

In case you want to use it as a base or take some ideas from there :)

**Comment by erikyo** [comment]
@mitogh want to share with you a plugin (never published) that has maybe some function that you might be interested in....

This is the filter with which I analyze in content and replace images, additionally adds a "count" to each image to let the user to decide how many image without lazyload:
https://github.com/erikyo/easylazy/blob/40e928a446449916ad10f41d10268b94eec949dc/webp-lazyload.php#L93

And this is the "companion" lazyload script that is used to serve the "best" image given the http image_accept header, (with IntersectionObserver and MutationObserver): 
https://github.com/erikyo/easylazy/blob/40e928a446449916ad10f41d10268b94eec949dc/webp-lazyload.php#L122

When i did this plugin i didn't want to modify the Wordpress media array and for this i created an exact "copy" of the image.(jpg|webp), but i guess with the current solution it would be even easier to "choose" the image with a regex

**Comment by mitogh** [comment]
Awesome thanks a lot for sharing your work with us @erikyo this is really helpful and would provide us with some ideas for sure. I will take a closer look.

`image_accept` is for sure a good improvement fur now I think the idea would be to not rely on it but as we move forward the idea would be to integrate it with some balance between server and client hints for detection.  More details here https://github.com/WordPress/performance/issues/149#issuecomment-1030292648

**Comment by erikyo** [comment]
About "image_accept" I know that there are some disadvantages (such as https://github.com/WordPress/performance/issues/149#issuecomment-1030192647 or even the old browser with a partial support for webp) but when I had chosen to use that way [it seemed to me the best method](https://developers.google.com/speed/webp/faq#server-side_content_negotiation_via_accept_headers) because the headers are available directly from the page request and without further implications. 

Using other methods might mean delaying the moment when images are served (which is not ideal given the purpose of this plugin). Putting advantages and disadvantages on the scales, however, I have no doubts to say that for me it is the best method.
